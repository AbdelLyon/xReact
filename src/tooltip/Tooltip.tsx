import { ReactNode } from "react";
import { TooltipProps as TooltipPropsRoot } from "@radix-ui/react-tooltip";

import {
  Tooltip as TooltipRoot,
  TooltipContent,
  TooltipProvider,
  TooltipTrigger,
} from "@/ui/tooltip";
import { cn } from "@/utils";

interface TooltipProps extends Omit<TooltipPropsRoot, "children"> {
  trigger: ReactNode;
  content: ReactNode;
  triggerAsChild?: boolean;
  delayDuration?: number;
  side?: "top" | "right" | "bottom" | "left";
  align?: "start" | "center" | "end";
  classNames?: {
    root?: string;
    trigger?: string;
    content?: string;
  };
  open?: boolean;
  onOpenChange?: (open: boolean) => void;
  sideOffset?: number;
}

export const Tooltip = ({
  trigger,
  content,
  triggerAsChild = false,
  delayDuration = 300,
  side = "top",
  align = "center",
  classNames = {},
  open,
  onOpenChange,
  sideOffset,
  ...props
}: TooltipProps) => {
  return (
    <TooltipProvider>
      <TooltipRoot
        delayDuration={delayDuration}
        open={open}
        onOpenChange={onOpenChange}
        {...props}
      >
        <TooltipTrigger asChild={triggerAsChild} className={classNames.trigger}>
          {trigger}
        </TooltipTrigger>
        <TooltipContent
          align={align}
          className={cn("z-50", classNames.content)}
          side={side}
          sideOffset={sideOffset}
        >
          {content}
        </TooltipContent>
      </TooltipRoot>
    </TooltipProvider>
  );
};
