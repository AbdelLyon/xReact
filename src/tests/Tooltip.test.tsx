import { HTMLAttributes, PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import { describe, it, expect, vi } from "vitest";

import { Tooltip } from "@/tooltip";

// Mock des dÃ©pendances
vi.mock("@/ui/tooltip", () => ({
  Tooltip: ({ children }: PropsWithChildren) => (
    <div data-testid="tooltip-root">{children}</div>
  ),
  TooltipContent: ({
    children,
    ...props
  }: PropsWithChildren<HTMLAttributes<HTMLDivElement>>) => (
    <div data-testid="tooltip-content" {...props}>
      {children}
    </div>
  ),
  TooltipProvider: ({ children }: PropsWithChildren) => (
    <div data-testid="tooltip-provider">{children}</div>
  ),
  TooltipTrigger: ({
    children,
    ...props
  }: PropsWithChildren<HTMLAttributes<HTMLDivElement>>) => (
    <div data-testid="tooltip-trigger" {...props}>
      {children}
    </div>
  ),
}));

vi.mock("@/utils", () => ({
  cn: (...args: any[]) => args.filter(Boolean).join(" "),
}));

describe("Tooltip Component", () => {
  it("renders correctly with default props", () => {
    const { asFragment } = render(
      <Tooltip
        content="This is a tooltip"
        trigger={<button>Hover me</button>}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correctly with custom props", () => {
    const { asFragment } = render(
      <Tooltip
        align="start"
        classNames={{
          trigger: "custom-trigger",
          content: "custom-content",
        }}
        content={<div>Custom content</div>}
        delayDuration={500}
        side="right"
        sideOffset={10}
        trigger={<span>Custom trigger</span>}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correctly with triggerAsChild prop", () => {
    const { asFragment } = render(
      <Tooltip
        content="Child content"
        trigger={<button>Child trigger</button>}
        triggerAsChild={true}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correctly with open and onOpenChange props", () => {
    const onOpenChange = vi.fn();
    const { asFragment } = render(
      <Tooltip
        content="Controlled content"
        open={true}
        trigger={<button>Controlled tooltip</button>}
        onOpenChange={onOpenChange}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
