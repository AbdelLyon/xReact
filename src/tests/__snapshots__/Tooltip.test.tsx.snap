// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tooltip Component > renders correctly with custom props 1`] = `
<DocumentFragment>
  <div
    data-testid="tooltip-provider"
  >
    <div
      data-testid="tooltip-root"
    >
      <div
        class="custom-trigger"
        data-testid="tooltip-trigger"
      >
        <span>
          Custom trigger
        </span>
      </div>
      <div
        align="start"
        class="z-50 custom-content"
        data-testid="tooltip-content"
        side="right"
        sideoffset="10"
      >
        <div>
          Custom content
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tooltip Component > renders correctly with default props 1`] = `
<DocumentFragment>
  <div
    data-testid="tooltip-provider"
  >
    <div
      data-testid="tooltip-root"
    >
      <div
        data-testid="tooltip-trigger"
      >
        <button>
          Hover me
        </button>
      </div>
      <div
        align="center"
        class="z-50"
        data-testid="tooltip-content"
        side="top"
      >
        This is a tooltip
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tooltip Component > renders correctly with open and onOpenChange props 1`] = `
<DocumentFragment>
  <div
    data-testid="tooltip-provider"
  >
    <div
      data-testid="tooltip-root"
    >
      <div
        data-testid="tooltip-trigger"
      >
        <button>
          Controlled tooltip
        </button>
      </div>
      <div
        align="center"
        class="z-50"
        data-testid="tooltip-content"
        side="top"
      >
        Controlled content
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tooltip Component > renders correctly with triggerAsChild prop 1`] = `
<DocumentFragment>
  <div
    data-testid="tooltip-provider"
  >
    <div
      data-testid="tooltip-root"
    >
      <div
        data-testid="tooltip-trigger"
      >
        <button>
          Child trigger
        </button>
      </div>
      <div
        align="center"
        class="z-50"
        data-testid="tooltip-content"
        side="top"
      >
        Child content
      </div>
    </div>
  </div>
</DocumentFragment>
`;
